src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -93,31 +93,31 @@ private static bool IsValueWriteAtomic()
             // See http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-335.pdf
             //
             Type valueType = typeof(TValue);
-            if (valueType.IsEnum)
+            if (!valueType.IsValueType)
             {
-                valueType = Enum.GetUnderlyingType(valueType);
+                return true;
             }
-            bool isAtomic =
-                !valueType.GetTypeInfo().IsValueType ||
-                valueType == typeof(bool) ||
-                valueType == typeof(char) ||
-                valueType == typeof(byte) ||
-                valueType == typeof(sbyte) ||
-                valueType == typeof(short) ||
-                valueType == typeof(ushort) ||
-                valueType == typeof(int) ||
-                valueType == typeof(uint) ||
-                valueType == typeof(float);
-
-            if (!isAtomic && IntPtr.Size == 8)
+
+            switch (Type.GetTypeCode(valueType))
             {
-                isAtomic =
-                    valueType == typeof(double) ||
-                    valueType == typeof(long) ||
-                    valueType == typeof(ulong);
+                case TypeCode.Boolean:
+                case TypeCode.Byte:
+                case TypeCode.Char:
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.SByte:
+                case TypeCode.Single:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                    return true;
+                case TypeCode.Int64:
+                case TypeCode.DateTime:
```

Comment:
I agree that it is not a good practice to take dependency on undocumented internal implementation details, but it is unlikely for DateTime implementation to change.

Is it common to use DateTime with ConcurrentDictionary? I do not think it is, thus I would not bother writing fragile code for it.

Accidently, somebody asked me pretty same question last week in the context of https://stackoverflow.com/questions/42258346/can-datetime-tear-in-a-64-bit-environment . `AreReadsAndWritesAreGuaranteedAtomic` method given as "answer" is wrong. It is not possible to infer whether the struct assignment is atomic by just looking at its size. It would need to inspect the struct fields using reflection.